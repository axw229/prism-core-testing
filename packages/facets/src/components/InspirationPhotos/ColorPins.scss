.pin__wrapper {
  touch-action: none;
  align-items: center;
  background: none;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  left: 0;
  line-height: 0.85em;
  overflow: hidden;
  outline: 0;
  position: absolute;
  top: 0;
  width: auto;
  z-index: 1;
  &--active {
    z-index: 100;
  }
  &__pin-add {
    align-items: center;
    border: 2px solid white;
    box-sizing: border-box;
    display: none;
    height: 3em;
    justify-content: center;
    width: 4.75em;
    order: 2;
    &--active {
      display: flex;
      z-index: 101;
    }
    &--left {
      order: unset;
    }
  }
  &__pin-add-icon {
    margin: 0 1em 0 0;
    &--left {
      margin: 0 0 0 1em;
    }
  }
  &__pin-add-button {
    outline: none;
    &:focus {
      outline-style: solid;
      outline-width: 2px;
      outline-color: var(getThemeColor('secondary'));
    }
  }
}

.pin__chip {
  align-items: center;
  border: 2px solid white;
  border-radius: 2em;
  box-sizing: border-box;
  cursor: pointer;
  display: flex;
  height: 2em;
  justify-content: center;
  line-height: 2em;
  order: 3;
  text-align: center;
  width: 2em;
  outline: none;
  &:focus {
    outline: none;
  }
  &--active {
    height: 3em;
    width: 3em;
    z-index: 2;
  }
}

.pin__name {
  display: block;
  font-weight: bold;
  line-height: 1em;
  &__wrapper {
    background-color: white;
    border: 1px solid white;
    box-sizing: border-box;
    color: #333333;
    display: none;
    font-size: 0.95em;
    line-height: 0.98;
    order: 1;
    opacity: 1;
    padding: 0.5em 0.9em;
    white-space: nowrap;
    width: auto;
    z-index: 100;
    &--active {
      display: block;
    }
    &--inactive {
      display: none;
      width: 0;
    }
  }
}

.pin__content {
  display: flex;
  position: absolute;
  right: 50%;
  top: 0;
  width: max-content;
  z-index: -1;
  &--left {
    left: 50%;
  }
  &__wrapper {
    height: 0;
    margin-left: 360px;
    position: relative;
  }
}

@include respond-below(xs) {
  .pin__wrapper {
    &__pin-add {
      width: auto;
    }

    &__pin-add-icon {
      margin: 0 0.5rem;
      &--left {
        margin: 0 0.5rem;
      }
    }
  }

  .pin__name {
    &__wrapper {
      padding: 0.5rem;
    }
  }

  .pin__content {
    right: -50px;
    top: 30px;
    z-index: 0;
    &--left {
      left: -50px;
    }
  }
}
