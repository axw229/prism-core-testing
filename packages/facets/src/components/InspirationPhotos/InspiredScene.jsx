// @flow
import React, { useContext } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import Prism, { ColorPin, ImageColorPicker } from '@prism/toolkit'
import ConfigurationContext, { type ConfigurationContextType } from 'src/contexts/ConfigurationContext/ConfigurationContext'
import { fullColorNumber, getContrastYIQ } from 'src/shared/helpers/ColorUtils'
import { add } from 'src/store/actions/live-palette'
import useColors from '../../shared/hooks/useColors'

type InspiredSceneProps = {
  data: { img: string, initPins: { x: number, y: number }[] }
}

const InspiredScene = ({ data: { img, initPins } }: InspiredSceneProps) => {
  // We call useColor to load the colors for cases where loadColor has not already been called.
  // we don't use the colors from what is returned bc the data type for unorderedColors (sting[]) is much thinner in the hook
  // than what is in redux (Color[])
  // We consume the colors from the redux store, since it's a bit redundant to recreate what we already have access to
  // eslint-disable-next-line no-unused-vars
  const [colors] = useColors()
  const unorderedColors = useSelector(store => store.colors?.unorderedColors)
  const dispatch = useDispatch()
  const livePaletteColors = useSelector((store) => store.lp.colors)
  const { brandKeyNumberSeparator, colorWall: { colorSwatch = {} } } = useContext<ConfigurationContextType>(ConfigurationContext)
  const { colorNumOnBottom = false, houseShaped = false } = colorSwatch
  const baseClass = `image-color-picker${colorNumOnBottom ? `--name-number${houseShaped ? '-house-shaped' : ''}` : '--number-name'}`

  const isColorAdded = ({ colorNumber }): boolean => livePaletteColors.some((c) => c.colorNumber === colorNumber)

  return (
    <Prism style={{ padding: '0 0.125rem' }}>
      {unorderedColors && (
        <ImageColorPicker
          colors={unorderedColors}
          imgSrc={img}
          initialPinLocations={initPins.map((pin) => ({ x: pin.x * 100, y: pin.y * 100 }))}
          pinRenderer={(props) => {
            return <ColorPin
              {...props}
              buttonContent={(color) => (
                <FontAwesomeIcon
                  aria-label={isColorAdded(color) ? 'added' : 'add'}
                  icon={isColorAdded(color) ? ['fa', 'check-circle'] : ['fal', 'plus-circle']}
                  style={{ color: getContrastYIQ(color.hex) }}
                />
              )}
              labelContent={(color) => (
                <div className={baseClass}>
                  <p className={`${baseClass}__number`}>{fullColorNumber(color.brandKey, color.colorNumber, brandKeyNumberSeparator)}</p>
                  <p className={`${baseClass}__name`}>{color.name}</p>
                </div>
              )}
              onColorAdded={(color) => dispatch(add(color))}
            />
          }}
          removeButtonContent={
            <FontAwesomeIcon aria-label='remove' icon={['fa', 'trash']} style={{ display: 'inline-block' }} />
          }
        />
      )}
    </Prism>
  )
}

export default InspiredScene
