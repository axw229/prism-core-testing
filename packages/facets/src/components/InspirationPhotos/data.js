// @flow
import tinycolor from '@ctrl/tinycolor'
import { type Color } from 'src/shared/types/Colors.js.flow'
import uniqueId from 'lodash/uniqueId'

export const throttleDragTime = 5
// we need this two const to make sure cursor always point to the center of preview circle
export const activedPinsHalfWidth = 24

export const cloneColorPinsArr = (pins) => {
  return pins.map(pin => ({ ...pin }))
}

export const getRGBInitPins = (acgColors: Array<Object>) => {
  return acgColors.map<Object>((acgColor: any): Object => {
    return {
      rgbArray: [acgColor.r, acgColor.g, acgColor.b],
      rgbValue: `rgb(${acgColor.r},${acgColor.g},${acgColor.b})`,
      translateValueX: acgColor.x,
      translateValueY: acgColor.y
    }
  })
}

export const rgb2lab = (rgb: Array<any>) => {
  let r = rgb[0] / 255
  let g = rgb[1] / 255
  let b = rgb[2] / 255
  let x; let y; let z

  r = (r > 0.04045) ? Math.pow((r + 0.055) / 1.055, 2.4) : r / 12.92
  g = (g > 0.04045) ? Math.pow((g + 0.055) / 1.055, 2.4) : g / 12.92
  b = (b > 0.04045) ? Math.pow((b + 0.055) / 1.055, 2.4) : b / 12.92

  x = (r * 0.4124 + g * 0.3576 + b * 0.1805) / 0.95047
  y = (r * 0.2126 + g * 0.7152 + b * 0.0722) / 1.00000
  z = (r * 0.0193 + g * 0.1192 + b * 0.9505) / 1.08883

  x = (x > 0.008856) ? Math.pow(x, 1 / 3) : (7.787 * x) + 16 / 116
  y = (y > 0.008856) ? Math.pow(y, 1 / 3) : (7.787 * y) + 16 / 116
  z = (z > 0.008856) ? Math.pow(z, 1 / 3) : (7.787 * z) + 16 / 116

  return [(116 * y) - 16, 500 * (x - y), 200 * (y - z)]
}

export type RenderingPin = {
  isActiveFlag: boolean,
  name: string,
  colorNumber: number,
  rgbValue: string,
  pinNumber: number,
  isContentLeft: boolean,
  translateX: number,
  translateValueY: number,
  pinId: string
}

export const renderingPins = (initPins: Array<Object>, canvasOffsetWidth: number, canvasOffsetHeight: number, colors: Color[]): RenderingPin[] => {
  return getRGBInitPins(initPins).map<Object>((acgColor: any, index: number): Object => {
    const color: Color = findClosestColor(acgColor.rgbArray, colors)
    const calculateTranslateX = acgColor.translateValueX * canvasOffsetWidth
    const calculateTranslateY = acgColor.translateValueY * canvasOffsetHeight

    return {
      ...color,
      isActiveFlag: (initPins.length - 1 === index),
      rgbValue: `rgb(${color.red},${color.green},${color.blue})`,
      pinNumber: index,
      pinId: uniqueId('pin_'),
      isContentLeft: calculateTranslateX < canvasOffsetWidth / 2,
      translateX: Math.min(canvasOffsetWidth - 2 * activedPinsHalfWidth, calculateTranslateX),
      translateY: Math.min(canvasOffsetHeight - 2 * activedPinsHalfWidth, calculateTranslateY)
    }
  })
}

export const findClosestColor = ([red, green, blue]: [number, number, number], colors: Color[]): Color => {
  const currentPixelHex: string = tinycolor(`rgb (${red}, ${green}, ${blue})`).toHex()

  return colors.reduce(([closestColor, smallestDistanceSoFar], color): [Color, number] => {
    const colorDistance: number = measureDistance(currentPixelHex, tinycolor(`rgb(${color.red},${color.green},${color.blue})`).toHex())
    return colorDistance < smallestDistanceSoFar ? [color, colorDistance] : [closestColor, smallestDistanceSoFar]
  }, [colors[0], 99999])[0]
}

export const measureDistance = (color1: string, color2: string): number => {
  const position1 = getChromaLuma(color1)
  const position2 = getChromaLuma(color2)
  const redDelta = position1[0] - position2[0]
  const greenDelta = position1[1] - position2[1]
  const lumaDelta = position1[2] - position2[2]
  const distance = redDelta * redDelta + greenDelta * greenDelta + lumaDelta * lumaDelta
  return distance
}

export const getChromaLuma = (rgbString: string) => {
  const color = tinycolor(rgbString)
  const luminosity = color.getLuminance()
  const colorRgb = color.toRgb()
  const red = colorRgb.r
  const green = colorRgb.g
  const blue = colorRgb.b
  const sum = red + green + blue
  let redChroma
  let greenChroma
  if (sum > 25) {
    redChroma = red / sum
    greenChroma = green / sum
  } else {
    redChroma = 1 / 3
    greenChroma = 1 / 3
  }

  return [redChroma, greenChroma, luminosity]
}
