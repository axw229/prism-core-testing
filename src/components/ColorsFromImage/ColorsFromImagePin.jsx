// @flow
import React, { PureComponent, Fragment } from 'react'

type Props = {
  pinType: string,
  onClickMethod: Function,
  // activePinIndex: number,
  isActiveFlag: boolean,
  transformValue: string,
  RGBstring: string,
  previewColorName: string,
  previewColorNumber: string
}

class ColorsFromImagePin extends PureComponent<Props> {
  render () {
    // eslint-disable-next-line one-var
    const { previewColorName, previewColorNumber, transformValue, RGBstring, onClickMethod, isActiveFlag, pinType } = this.props

    let isPreviewPin = false
    let additionalPinClass = ''
    let isActiveClass = ''

    if (pinType === 'preview') {
      isPreviewPin = true
      additionalPinClass = 'pin--preview__wrapper'
    }

    if (isPreviewPin) {
      if (isActiveFlag) isActiveClass = 'pin--active__wrapper'
    } else {
      // TODO: fix this functionality
      // is `this.key` supposed to be a reference to the key prop? is it supposed to be internally tracked?
      // if (activePinIndex === this.key) isActiveClass = 'pin--active__wrapper'
    }

    return (
      <Fragment>
        <button onClick={onClickMethod} style={{ transform: transformValue }} className={additionalPinClass + ' pin__wrapper ' + isActiveClass}>
          <span className='pin__chip' style={{ background: RGBstring }} />
          <div className='pin__name-wrapper'>
            <span className='pin__copy pin__name'>{previewColorName}</span>
            <span className='pin__copy pin__number'>SW {previewColorNumber}</span>
          </div>
        </button>
      </Fragment>
    )
  }
}

export default ColorsFromImagePin
