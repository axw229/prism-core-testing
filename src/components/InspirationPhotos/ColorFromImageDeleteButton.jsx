// @flow

import React from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { injectIntl } from 'react-intl'
import 'src/providers/fontawesome/fontawesome'

type ColorFromImageDeleteButtonProps = {
  isVisible: boolean,
  clickHandler: Function,
  intl: any
}

const ColorFromImageDeleteButton = (props: ColorFromImageDeleteButtonProps) => {
  return (
    props.isVisible ? <button
      className='scene__image__wrapper__delete-pin'
      onClick={props.clickHandler}>
      <FontAwesomeIcon
        title={props.intl.formatMessage({ id: 'DELETE.PIN' })}
        icon='trash' size='1x' />
    </button> : null
  )
}

export default injectIntl(ColorFromImageDeleteButton)
