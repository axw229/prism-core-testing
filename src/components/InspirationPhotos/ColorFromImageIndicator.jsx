// @flow
import React from 'react'

type ColorFromImageIndicatorProps = {
  currentPixelRGBstring: string,
  top: number,
  bottom: number,
  left: number,
  right: number
}

const ColorFromImageIndicator = (props: ColorFromImageIndicatorProps) => {
  return (
    <div className='scene__image__wrapper__indicator'
      style={{
        backgroundColor: props.currentPixelRGBstring,
        top: props.top,
        bottom: props.bottom,
        left: props.left,
        right: props.right
      }} />
  )
}

export default ColorFromImageIndicator
