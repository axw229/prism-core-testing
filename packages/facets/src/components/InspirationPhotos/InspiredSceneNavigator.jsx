// @flow
import React, { useContext,useEffect, useState } from 'react'
import { useIntl } from 'react-intl'
import { shallowEqual,useDispatch, useSelector } from 'react-redux'
import CardMenu from 'src/components/CardMenu/CardMenu'
import ConfigurationContext, { type ConfigurationContextType } from 'src/contexts/ConfigurationContext/ConfigurationContext'
import { loadInspirationalPhotos } from '../../store/actions/inspirationalPhotos'
import Carousel from '../Carousel/Carousel'
import ColorCollectionsTab from '../Shared/ColorCollectionsTab'
import InspiredScene from './InspiredScene'
import '../ColorCollections/ColorCollections.scss'

const baseClass = 'color-collections'

const InspiredSceneNavigator = () => {
  const dispatch = useDispatch()
  const { formatMessage, locale } = useIntl()
  const { brandId, cvw = {} } = useContext<ConfigurationContextType>(ConfigurationContext)
  const { inspirationalPhotos } = cvw

  useEffect(() => { loadInspirationalPhotos(brandId, { language: locale })(dispatch) }, [])

  const { collectionTabs, flatData, tabMap } = useSelector(({ inspirationalPhotos: { data } }) => ({
    collectionTabs: data.flatMap((scene, i) => ({ id: `tab${i}`, tabName: scene.name })),
    tabMap: data.flatMap((scene, i) => Array(scene.sceneDefinition.length).fill(`tab${i}`)),
    flatData: data.flatMap((scene, i) => (
      scene.sceneDefinition.map((def, x) => ({ img: def.photoUrl, id: `tab${i}-${x}`, initPins: def.initPins }))
    ))
  }), shallowEqual)

  const [tabId: string, setTabId: string => void] = useState('tab0')

  return (
    <CardMenu menuTitle={inspirationalPhotos?.title ?? formatMessage({ id: 'INSPIRATIONAL_PHOTOS' })}>
      {() => (
        <div className={`${baseClass}__wrapper`}>
          <ColorCollectionsTab collectionsSelectLabel={inspirationalPhotos?.collectionsSelectLabel} collectionTabs={collectionTabs} tabIdShow={tabId} showTab={setTabId} />
          <div className={`${baseClass}__collections-list`} role='main'>
            <Carousel
              BaseComponent={InspiredScene}
              data={flatData}
              defaultItemsPerView={1}
              tabId={tabId}
              setTabId={setTabId}
              isInfinity
              tabMap={tabMap}
            />
          </div>
        </div>
      )}
    </CardMenu>
  )
}

export default InspiredSceneNavigator
