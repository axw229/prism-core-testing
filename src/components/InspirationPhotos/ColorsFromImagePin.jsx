// @flow
import React, { useState, useRef, useEffect } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { useIntl } from 'react-intl'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { add } from '../../store/actions/live-palette'
import type { Color } from 'src/shared/types/Colors.js.flow'
import { LiveMessage } from 'react-aria-live'
import { activedPinsHalfWidth } from './data'
import { getContrastYIQ } from '../../../src/shared/helpers/ColorUtils'
import { KEY_CODES } from 'src/constants/globals'
import './ColorPins.scss'
import 'src/scss/convenience/visually-hidden.scss'
import 'src/providers/fontawesome/fontawesome'

type Props = {
  color: {
    ...Color,
    isActiveFlag: boolean,
    isContentLeft: boolean,
    translateX: number,
    translateY: number,
    rgbValue: string,
    pinNumber: number,
    isMovingPin: boolean,
    hide: boolean,
    pinId: string
  },
  useTranslateProps: boolean,
  translateX: number,
  translateY: number,
  activatePin: Function,
  handleDrag: Function,
  deleteCurrentPin: Function,
  handleDragStop: Function,
  handlePinMoveByKeyboard: Function,
  handleKeyUpAfterPinMove: Function,
  handleTouchEnd: Function
}

export const CLASSES = { BASE: 'pin__wrapper', CHECKBOX: 'visually-hidden' }

const PIN_MOVEMENT_INTERVAL = 10
const PIN_MOVEMENT_SHIFT_KEY_INTERVAL = 1

export default (props: Props) => {
  const { color, activatePin, handlePinMoveByKeyboard, handleKeyUpAfterPinMove, deleteCurrentPin, handleDrag, handleDragStop, handleTouchEnd } = props

  const dispatch = useDispatch()
  const { formatMessage } = useIntl()
  const livePaletteColors = useSelector(store => store.lp.colors)
  const colorDivRef: { current?: HTMLDivElement } = useRef()
  const [isColorDivFocused: boolean, setIsColorDivFocused: boolean => void] = useState()
  const isColorAdded = livePaletteColors.some(paletteColor => paletteColor.colorNumber === color.colorNumber)
  const [isHidden, setIsHidden] = useState(false)

  useEffect(() => {
    setIsHidden(color.hide)
  }, [color])

  const focusHandler = () => {
    colorDivRef.current && colorDivRef.current.focus()
    if (isColorDivFocused && colorDivRef.current) {
      colorDivRef.current.style.outlineStyle = 'solid'
      colorDivRef.current.style.outlineWidth = '2px'
      colorDivRef.current.style.outlineColor = '#2cabe2'
    }
  }

  const getElementsTopLeftCoordinates = (elem) => {
    const { top, left } = elem.target.getBoundingClientRect()
    return { top, left }
  }

  const onKeyDownHandlerDiv = (e) => {
    e.keyCode === KEY_CODES.KEY_CODE_TAB && setIsColorDivFocused(true)

    if (isColorDivFocused && (e.keyCode === KEY_CODES.KEY_CODE_ENTER || e.keyCode === KEY_CODES.KEY_CODE_SPACE)) {
      e.stopPropagation()
      e.preventDefault()
      activatePin(color.pinNumber)
      focusHandler()
    } else if (isColorDivFocused &&
      color.isActiveFlag && (
      e.keyCode === KEY_CODES.KEY_CODE_ARROW_LEFT ||
      e.keyCode === KEY_CODES.KEY_CODE_ARROW_UP ||
      e.keyCode === KEY_CODES.KEY_CODE_ARROW_RIGHT ||
      e.keyCode === KEY_CODES.KEY_CODE_ARROW_DOWN)
    ) {
      e.stopPropagation()
      e.preventDefault()
      const pinPositionDocument = getElementsTopLeftCoordinates(e)
      const movementInterval = e.shiftKey ? PIN_MOVEMENT_SHIFT_KEY_INTERVAL : PIN_MOVEMENT_INTERVAL
      let pinTopReferredToCanvas = Math.floor(pinPositionDocument.top) + activedPinsHalfWidth
      let pinLeftReferredToCanvas = Math.floor(pinPositionDocument.left) + activedPinsHalfWidth

      switch (e.keyCode) {
        case KEY_CODES.KEY_CODE_ARROW_LEFT:
          pinLeftReferredToCanvas -= movementInterval
          break
        case KEY_CODES.KEY_CODE_ARROW_UP:
          pinTopReferredToCanvas -= movementInterval
          break
        case KEY_CODES.KEY_CODE_ARROW_RIGHT:
          pinLeftReferredToCanvas += movementInterval
          break
        case KEY_CODES.KEY_CODE_ARROW_DOWN:
          pinTopReferredToCanvas += movementInterval
          break
      }

      handlePinMoveByKeyboard({ offsetX: pinLeftReferredToCanvas, offsetY: pinTopReferredToCanvas, pinNumber: color.pinNumber })
    }
  }

  const handleKeyUpDiv = (e) => {
    if (e.keyCode === KEY_CODES.KEY_CODE_ENTER || e.keyCode === KEY_CODES.KEY_CODE_SPACE) {
      focusHandler()
    } else if (color.isActiveFlag && (
      e.keyCode === KEY_CODES.KEY_CODE_ARROW_LEFT ||
      e.keyCode === KEY_CODES.KEY_CODE_ARROW_UP ||
      e.keyCode === KEY_CODES.KEY_CODE_ARROW_RIGHT ||
      e.keyCode === KEY_CODES.KEY_CODE_ARROW_DOWN
    )) {
      handleKeyUpAfterPinMove(color.pinNumber)
    }
  }

  const handleMouseMove = (e: SyntheticEvent, isMobile: booleam = false) => {
    if (e.target) {
      e.stopPropagation()
    }

    handleDrag(e.clientX, e.clientY, isMobile)
  }

  const handlePinDragStop = (e: SyntheticEvent) => {
    e.stopPropagation()
    handleDragStop(e)
    window.removeEventListener('mousemove', handleMouseMove)
    window.removeEventListener('mouseup', handlePinDragStop)
  }

  const handlePinDragStart = (e) => {
    deleteCurrentPin(color.pinNumber)
    window.addEventListener('mousemove', handleMouseMove)
    window.addEventListener('mouseup', handlePinDragStop)
  }

  const handleTouchStart = (e: SyntheticEvent) => {
    setIsColorDivFocused(false)
    setIsHidden(true)
  }

  const handleTouchStop = (e: SyntheticEvent) => {
    const dims = colorDivRef.current.getBoundingClientRect()
    handleTouchEnd(dims, color.pinId)
  }

  const handleTouchMove = (e: SyntheticEvent) => {
    // @todo this gets drag working on mobile but color is off, this is index math issue.  I should build a translator...
    handleMouseMove({ clientX: e.targetTouches[0].clientX, clientY: e.targetTouches[0].clientY }, true)
  }

  const translateX = props.useTranslateProps ? props.translateX : color.translateX
  const translateY = props.useTranslateProps ? props.translateY : color.translateY
  return (
    <>
      <label
        title={color.name}
        aria-label={color.name}
        htmlFor={color.name}
        tabIndex='-1'
        onClick={(e) => {
          e.preventDefault()
          e.stopPropagation()
          activatePin(color.pinNumber)
        }}
        style={{ transform: `translate(${translateX - 360}px, ${translateY}px)`, overflow: 'visible', display: isHidden ? 'none' : 'flex' }}
        className={`pin__wrapper ${color.isActiveFlag ? 'pin__wrapper--active' : ''}`}
      >
        <div className='pin__content__wrapper'>
          <div
            ref={colorDivRef}
            role='button'
            draggable
            tabIndex='0'
            onKeyDown={onKeyDownHandlerDiv}
            onKeyUp={handleKeyUpDiv}
            onFocus={focusHandler}
            onBlur={() => {
              colorDivRef.current.blur()
              colorDivRef.current.style.outline = 'none'
            }}
            onMouseDown={() => setIsColorDivFocused(false)}
            onDragStart={handlePinDragStart}
            onTouchStart={handleTouchStart}
            onTouchMove={handleTouchMove}
            onTouchEnd={handleTouchStop}
            className={`pin__chip ${(color.isActiveFlag) ? `pin__chip--active` : ''}`}
            style={{ background: color.rgbValue }}
          >
            <input
              tabIndex='-1'
              className={CLASSES.CHECKBOX}
              type='checkbox'
              checked={color.isActiveFlag}
              name={`pin-${color.pinNumber}`}
              id={`pin-${color.pinNumber}`}
              onChange={() => activatePin(color.pinNumber)}
            />
            {(isColorAdded && !color.isActiveFlag) && (<FontAwesomeIcon style={{ color: getContrastYIQ(color.hex) }} icon={['fa', 'check-circle']} size='1x' />)}
          </div>
          <div className={`pin__content ${color.isContentLeft ? 'pin__content--left' : ''}`}>
            <div className={`pin__name__wrapper ${color.isActiveFlag ? `pin__name__wrapper--active` : `pin__name__wrapper--inactive`}`}>
              <span className='pin__copy pin__number'>{(color.colorNumber) ? `${color.brandKey}${color.colorNumber}` : ''}</span>
              <span className='pin__copy pin__name'>{color.name}</span>
            </div>
            <div
              className={`pin__wrapper__pin-add ${color.isContentLeft ? 'pin__wrapper__pin-add--left' : ''} ${color.isActiveFlag ? `pin__wrapper__pin-add--active` : ''}`}
              style={{ background: color.rgbValue }}
            >
              <div className={`pin__wrapper__pin-add-icon ${color.isContentLeft ? 'pin__wrapper__pin-add-icon--left' : ''}`}>
                {(isColorAdded)
                  ? (<FontAwesomeIcon style={{ color: getContrastYIQ(color.hex) }} icon={['fa', 'check-circle']} size='2x' />)
                  : (
                    <button
                      tabIndex={(color.isActiveFlag) ? `0` : `-1`}
                      className={`pin__wrapper__pin-add-button`}
                      onClick={(e) => {
                        e.stopPropagation()
                        dispatch(add(color))
                      }}
                    >
                      <FontAwesomeIcon tabIndex='-1' style={{ color: getContrastYIQ(color.hex) }} icon={['fal', 'plus-circle']} size='2x' />
                    </button>
                  )
                }
              </div>
            </div>
          </div>
        </div>
      </label>
      {color.isActiveFlag && <LiveMessage message={formatMessage({ id: 'LIVE_MESSAGE_COLOR_NAME' }, { previewColorName: color.colorName })} aria-live='polite' />}
      {color.isActiveFlag && isColorDivFocused && <LiveMessage message={formatMessage({ id: 'LIVE_MESSAGE_MOVE_PIN' })} aria-live='assertive' clearOnUnmount='true' />}
    </>
  )
}
