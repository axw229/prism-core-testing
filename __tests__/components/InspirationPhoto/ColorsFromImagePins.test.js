import React from 'react'
import { shallow } from 'enzyme'
import { ColorsFromImagePin } from 'src/components/InspirationPhotos/ColorsFromImagePin'
import { LiveMessage } from 'react-aria-live'

const mockFn = jest.fn()
let defaultProps = {
  activedPins: mockFn,
  isActiveFlag: true,
  isContentLeft: true,
  translateX: 0,
  translateY: 0,
  RGBstring: '',
  previewColorName: '',
  previewColorNumber: '0001',
  pinNumber: 0,
  addColors: [],
  addToLivePalette: mockFn,
  handleDrag: mockFn,
  deleteCurrentPin: mockFn,
  handleDragStop: mockFn,
  handlePinMoveByKeyboard: mockFn,
  handleKeyUpAfterPinMove: mockFn,
  hide: true,
  isMovingPin: true
}

const createColorsFromImagePin = (props) => {
  let newProps = Object.assign({}, defaultProps, props)
  return shallow(<ColorsFromImagePin {...newProps} />, { disableLifecycleMethods: true })
}

const wrapper = createColorsFromImagePin()
const drag = jest.fn()
const mouseDown = jest.fn()
const blur = jest.fn()
const focus = jest.fn()
const keyUp = jest.fn()
const button = `[role='button']`

describe('Testing state for createColorsFromImage Pins component', () => {
  it('State should initialize correctly', () => {
    const initialState = wrapper.instance().state
    expect(initialState).toEqual({
      isColorDivFocused: true
    })
  })
})

describe('compoments rendring test', () => {
  it('should rendering ColorsFromImagePin Component when it active', () => {
    expect(wrapper.find('.pin__wrapper--active').exists()).toBe(true)
  })

  it('should rendering Colors From Image pins', () => {
    expect(wrapper.find('.pin__content__wrapper').exists()).toBe(true)
  })

  it('should rendering LiveMessage', () => {
    const previewColorName = wrapper.instance().props.previewColorName
    const component = wrapper.find(LiveMessage).at(0)
    expect(component.prop('message')).toEqual(`Color ${previewColorName} is active. Press tab to go to other color pins`)
  })

  it('should rendering LiveMessage focus Message', () => {
    const component = wrapper.find(LiveMessage).at(1)
    expect(component.prop('message')).toEqual('Use the arrow keys on your keyboard to move your picked color. Hold shift for fine movement.')
  })
})

describe('component event test', () => {
  it('click pin should works fine', () => {
    const click = jest.fn()
    wrapper.instance().onClickHandlerLabel = click
    wrapper.instance().forceUpdate()
    wrapper.find('.pin__wrapper--active').simulate('click')
    expect(click).toHaveBeenCalled()
  })

  it('pins should be dragable', () => {
    wrapper.instance().handleDragStart = drag
    wrapper.instance().forceUpdate()
    wrapper.find(button).simulate('dragStart')
    expect(drag).toHaveBeenCalled()
  })

  it('pins should be clickable', () => {
    wrapper.instance().handleMouseDown = mouseDown
    wrapper.instance().forceUpdate()
    wrapper.find(button).simulate('mouseDown')
    expect(mouseDown).toHaveBeenCalled()
  })

  it('pins should be handled by keyboard', () => {
    wrapper.instance().handleKeyUpDiv = keyUp
    wrapper.instance().forceUpdate()
    wrapper.find(button).simulate('keyUp')
    expect(keyUp).toHaveBeenCalled()
  })

  it('pins could be blur', () => {
    wrapper.instance().blurHandler = blur
    wrapper.instance().forceUpdate()
    wrapper.find(button).simulate('blur')
    expect(blur).toHaveBeenCalled()
  })

  it('pins could be focus', () => {
    wrapper.instance().focusHandler = focus
    wrapper.instance().forceUpdate()
    wrapper.find(button).simulate('focus')
    expect(focus).toHaveBeenCalled()
  })
})
