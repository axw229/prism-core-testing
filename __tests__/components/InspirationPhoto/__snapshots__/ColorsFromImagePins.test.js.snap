// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot match testing snapshot testing 1`] = `
<Fragment>
  <label
    aria-label=""
    className="pin__wrapper pin__wrapper--active"
    htmlFor=""
    onClick={[Function]}
    style={
      Object {
        "display": "none",
        "overflow": "visible",
        "transform": "translate(-360px, 0px)",
      }
    }
    tabIndex="-1"
    title=""
  >
    <div
      className="pin__content__wrapper"
    >
      <div
        className="pin__chip pin__chip--active"
        draggable={true}
        onBlur={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        role="button"
        style={
          Object {
            "background": "",
          }
        }
        tabIndex="0"
      >
        <input
          checked={true}
          className="visually-hidden"
          id="pin-0"
          name="pin-0"
          onChange={[Function]}
          tabIndex="-1"
          type="checkbox"
        />
      </div>
      <div
        className="pin__content pin__content--left"
      >
        <div
          className="pin__name__wrapper pin__name__wrapper--active"
        >
          <span
            className="pin__copy pin__number"
          >
            SW0001
          </span>
          <span
            className="pin__copy pin__name"
          />
        </div>
        <div
          className="pin__wrapper__pin-add pin__wrapper__pin-add--left pin__wrapper__pin-add--active"
          style={
            Object {
              "background": "",
            }
          }
        >
          <div
            className="pin__wrapper__pin-add-icon pin__wrapper__pin-add-icon--left"
          >
            <button
              className="pin__wrapper__pin-add-button"
              onClick={[Function]}
              tabIndex="0"
            >
              <FontAwesomeIcon
                border={false}
                className=""
                fixedWidth={false}
                flip={null}
                icon={
                  Array [
                    "fal",
                    "plus-circle",
                  ]
                }
                inverse={false}
                listItem={false}
                mask={null}
                pull={null}
                pulse={false}
                rotation={null}
                size="2x"
                spin={false}
                style={
                  Object {
                    "color": "#000",
                  }
                }
                symbol={false}
                tabIndex="-1"
                title=""
                transform={null}
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </label>
  <LiveMessage
    aria-live="polite"
    message="Color  is active. Press tab to go to other color pins"
  />
  <LiveMessage
    aria-live="assertive"
    clearOnUnmount="true"
    message="Use the arrow keys on your keyboard to move your picked color. Hold shift for fine movement."
  />
</Fragment>
`;
